internal/docstore/driver/codec.go

```diff
@@ -0,0 +1,701 @@
+// Copyright 2019 The Go Cloud Development Kit Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// TODO(jba): support struct tags.
+// TODO(jba): for efficiency, enable encoding of only a subset of field paths.
+
+package driver
+
+import (
+	"encoding"
+	"fmt"
+	"reflect"
+	"strconv"
+
+	"github.com/golang/protobuf/proto"
+	"gocloud.dev/internal/docstore/internal/fields"
+	"gocloud.dev/internal/gcerr"
+)
+
+var (
+	binaryMarshalerType   = reflect.TypeOf((*encoding.BinaryMarshaler)(nil)).Elem()
+	binaryUnmarshalerType = reflect.TypeOf((*encoding.BinaryUnmarshaler)(nil)).Elem()
+	textMarshalerType     = reflect.TypeOf((*encoding.TextMarshaler)(nil)).Elem()
+	textUnmarshalerType   = reflect.TypeOf((*encoding.TextUnmarshaler)(nil)).Elem()
+	protoMessageType      = reflect.TypeOf((*proto.Message)(nil)).Elem()
+)
+
+// An Encoder encodes Go values in some other form (e.g. JSON, protocol buffers).
+// The encoding protocol is designed to avoid losing type information by passing
+// values using interface{}. An Encoder is responsible for storing the value
+// it is encoding.
+//
+// Because all providers must support the same set of values, the encoding methods
+// (with the exception of EncodeStruct) do not return errors. EncodeStruct is special
+// because it is an escape hatch for arbitrary structs, not all of which may be
+// encodable.
+type Encoder interface {
+	// These methods all encode and store a single Go value.
+	EncodeNull()
+	EncodeBool(bool)
+	EncodeString(string)
+	EncodeInt(int64)
+	EncodeUint(uint64)
+	EncodeFloat(float64)
+	EncodeComplex(complex128)
```

Comment:
Nit: consider dropping the `Encode` part of all these function names?

Reply:
Seems weird to have methods named `Int`, `Float`, etc., even if they are in a type called `FooEncoder`.

Leaving it for now, will change if it continues to bother you.